README.txt
Introduction
============
This package contains utilities used to package some of STScI's Python
projects; specifically those projects that comprise stsci_python_ and
Astrolib_.

It currently consists mostly of some setup_hook scripts meant for use with
`distutils2/packaging`_ and/or d2to1_, and a customized easy_install command
meant for use with distribute_.

This package is not meant for general consumption, though it might be worth
looking at for examples of how to do certain things with your own packages, but
YMMV.

Features
========

Hook Scripts
------------
Currently the main features of this package are a couple of setup_hook scripts.
In distutils2, a setup_hook is a script that runs at the beginning of any
pysetup command, and can modify the package configuration read from setup.cfg.
There are also pre- and post-command hooks that only run before/after a
specific setup command (eg. build_ext, install) is run.

stsci.distutils.hooks.use_packages_root
'''''''''''''''''''''''''''''''''''''''
If using the ``packages_root`` option under the ``[files]`` section of
setup.cfg, this hook will add that path to ``sys.path`` so that modules in your
package can be imported and used in setup.  This can be used even if
``packages_root`` is not specified--in this case it adds ``''`` to
``sys.path``.

stsci.distutils.hooks.version_setup_hook
''''''''''''''''''''''''''''''''''''''''
Creates a Python module called version.py which currently contains four
variables:

* ``__version__`` (the release version)
* ``__svn_revision__`` (the SVN revision info as returned by the ``svnversion``
  command)
* ``__svn_full_info__`` (as returned by the ``svn info`` command)
* ``__setup_datetime__`` (the date and time that setup.py was last run).

These variables can be imported in the package's `__init__.py` for degugging
purposes.  The version.py module will *only* be created in a package that
imports from the version module in its `__init__.py`.  It should be noted that
this is generally preferable to writing these variables directly into
`__init__.py`, since this provides more control and is less likely to
unexpectedly break things in `__init__.py`.

stsci.distutils.hooks.version_pre_command_hook
''''''''''''''''''''''''''''''''''''''''''''''
Identical to version_setup_hook, but designed to be used as a pre-command
hook.

stsci.distutils.hooks.version_post_command_hook
'''''''''''''''''''''''''''''''''''''''''''''''
The complement to version_pre_command_hook.  This will delete any version.py
files created during a build in order to prevent them from cluttering an SVN
working copy (note, however, that version.py is *not* deleted from the build/
directory, so a copy of it is still preserved).  It will also not be deleted
if the current directory is not an SVN working copy.  For example, if source
code extracted from a source tarball it will be preserved.

stsci.distutils.hooks.tag_svn_revision
''''''''''''''''''''''''''''''''''''''
A setup_hook to add the SVN revision of the current working copy path to the
package version string, but only if the version ends in .dev.

For example, ``mypackage-1.0.dev`` becomes ``mypackage-1.0.dev1234``.  This is
in accordance with the version string format standardized by PEP 386.

This should be used as a replacement for the ``tag_svn_revision`` option to
the egg_info command.  This hook is more compatible with packaging/distutils2,
which does not include any VCS support.  This hook is also more flexible in
that it turns the revision number on/off depending on the presence of ``.dev``
in the version string, so that it's not automatically added to the version in
final releases.

This hook does require the ``svnversion`` command to be available in order to
work.  It does not examine the working copy metadata directly.

stsci.distutils.hooks.numpy_extension_hook
''''''''''''''''''''''''''''''''''''''''''
This is a pre-command hook for the build_ext command.  To use it, add a
``[build_ext]`` section to your setup.cfg, and add to it::

    pre-hook.numpy-extension-hook = stsci.distutils.hooks.numpy_extension_hook

This hook must be used to build extension modules that use Numpy.   The primary
side-effect of this hook is to add the correct numpy include directories to
`include_dirs`.  To use it, add 'numpy' to the 'include-dirs' option of each
extension module that requires numpy to build.  The value 'numpy' will be
replaced with the actual path to the numpy includes.

stsci.distutils.hooks.is_display_option
'''''''''''''''''''''''''''''''''''''''
This is not actually a hook, but is a useful utility function that can be used
in writing other hooks.  Basically, it returns ``True`` if setup.py was run
with a "display option" such as --version or --help.  This can be used to
prevent your hook from running in such cases.

stsci.distutils.hooks.glob_data_files
'''''''''''''''''''''''''''''''''''''
A pre-command hook for the install_data command.  Allows filename wildcards as
understood by ``glob.glob()`` to be used in the data_files option.  This hook
must be used in order to have this functionality since it does not normally
exist in distutils.

This hook also ensures that data files are installed relative to the package
path.  data_files shouldn't normally be installed this way, but the
functionality is required for a few special cases.


Commands
--------
build_optional_ext
''''''''''''''''''
This serves as an optional replacement for the default built_ext command,
which compiles C extension modules.  Its purpose is to allow extension modules
to be *optional*, so that if their build fails the rest of the package is
still allowed to be built and installed.  This can be used when an extension
module is not definitely required to use the package.

To use this custom command, add::

    commands = stsci.distutils.command.build_optional_ext.build_optional_ext

under the ``[global]`` section of your package's setup.cfg.  Then, to mark
an individual extension module as optional, under the setup.cfg section for
that extension add::

    optional = True

Optionally, you may also add a custom failure message by adding::

    fail_message = The foobar extension module failed to compile: %(message)s.
                   This could be because you lack such and such headers.
                   This package will still work, but such and such features
                   will be disabled.

Note that if the string ``%message`` is present in the ``fail_message``, it
will be replaced with whatever error caused the build to fail.


.. _stsci_python: http://www.stsci.edu/resources/software_hardware/pyraf/stsci_python
.. _Astrolib: http://www.scipy.org/AstroLib/
.. _distutils2/packaging: http://distutils2.notmyidea.org/
.. _d2to1: http://pypi.python.org/pypi/d2to1
.. _distribute: http://pypi.python.org/pypi/distribute


CHANGES.txt
Changelog
===========

0.3.7 (2013-12-23)
------------------

- Avoid using ``Popen.stdout`` directly in the version.py SVN revision
  auto-update script to avoid possible ResourceWarnings on Python >= 3.2.
  See https://github.com/spacetelescope/PyFITS/issues/45


0.3.6 (2013-11-21)
------------------

- Fixed a syntax error in Python 3 that was introduced in 0.3.5.  This
  could occur very early in the setup such that it bailed before even 2to3
  could run on the rest of the package.


0.3.5 (2013-11-18)
------------------

- Fixed an obscure issue that could occur when trying to install with
  easy_install on Python 2 systems that have lib2to3 installed but have never
  used it.


0.3.4 (2013-07-31)
------------------

- Updated the check for ``__loader__`` added in v0.3.3 to only perform
  that check on Python >= 3.3, since the issue doesn't apply to older
  Python versions.


0.3.3 (2013-07-25)
------------------

- Updated the import-time SVN revision update mechanism in the ``version.py``
  module generated by the ``version_setup_hook`` to avoid running when not in
  a dev version of the package.  This saves time on importing released
  packages when installed on users' systems.

- Added a workaround to a bug on Python 3.3 that could cause stsci.distutils
  to crash during installation.


0.3.2 (2013-03-27)
------------------

- Fixed a bug in the version hook that could occur if the svnversion command
  fails.

- Updated the template for the version.py module generated by the version hook
  so that ``from .version import *`` will work for applications.

- Added a ``__vdate__`` variable in version.py which may contain a release
  date string by specifying a ``vdate`` option in the ``[metadata]`` section
  of setup.cfg.

- Added a ``stsci_distutils_version`` variable in version.py containing the
  version of stsci.distutils used to generate the file--useful primarily for
  debugging purposes.

- Version 0.3.1 added a new zest.releaser hooks to ensure that source
  distributes are created as tar.gz files instead of zip files--this was left
  out of the changelog for 0.3.1.

- The tar.gz zest.releaser hook is updated in 0.3.2 to only run on stsci
  packages.


0.3.1 (2012-06-28)
------------------

- Fixed a bug where console output from svn-related programs was assumed to be
  ascii, leading to possible crashes on non-English systems.


0.3 (2012-02-20)
----------------

- The ``glob_data_files`` hook became a pre-command hook for the install_data
  command instead of being a setup-hook.  This is to support the additional
  functionality of requiring data_files with relative destination paths to be
  install relative to the package's install path (i.e. site-packages).

- Dropped support for and deprecated the easier_install custom command.
  Although it should still work, it probably won't be used anymore for
  stsci_python packages.

- Added support for the ``build_optional_ext`` command, which replaces/extends
  the default ``build_ext`` command.  See the README for more details.

- Added the ``tag_svn_revision`` setup_hook as a replacement for the
  setuptools-specific tag_svn_revision option to the egg_info command.  This
  new hook is easier to use than the old tag_svn_revision option: It's
  automatically enabled by the presence of ``.dev`` in the version string, and
  disabled otherwise.

- The ``svn_info_pre_hook`` and ``svn_info_post_hook`` have been replaced with
  ``version_pre_command_hook`` and ``version_post_command_hook`` respectively.
  However, a new ``version_setup_hook``, which has the same purpose, has been
  added.  It is generally easier to use and will give more consistent results
  in that it will run every time setup.py is run, regardless of which command
  is used.  ``stsci.distutils`` itself uses this hook--see the `setup.cfg` file
  and `stsci/distutils/__init__.py` for example usage.

- Instead of creating an `svninfo.py` module, the new ``version_`` hooks create
  a file called `version.py`.  In addition to the SVN info that was included
  in `svninfo.py`, it includes a ``__version__`` variable to be used by the
  package's `__init__.py`.  This allows there to be a hard-coded
  ``__version__`` variable included in the source code, rather than using
  pkg_resources to get the version.

- In `version.py`, the variables previously named ``__svn_version__`` and
  ``__full_svn_info__`` are now named ``__svn_revision__`` and
  ``__svn_full_info__``.

- Fixed a bug when using stsci.distutils in the installation of other packages
  in the ``stsci.*`` namespace package.  If stsci.distutils was not already
  installed, and was downloaded automatically by distribute through the
  setup_requires option, then ``stsci.distutils`` would fail to import.  This
  is because the way the namespace package (nspkg) mechanism currently works,
  all packages belonging to the nspkg *must* be on the import path at initial
  import time.

  So when installing stsci.tools, for example, if ``stsci.tools`` is imported
  from within the source code at install time, but before ``stsci.distutils``
  is downloaded and added to the path, the ``stsci`` package is already
  imported and can't be extended to include the path of ``stsci.distutils``
  after the fact.  The easiest way of dealing with this, it seems, is to
  delete ``stsci`` from ``sys.modules``, which forces it to be reimported, now
  the its ``__path__`` extended to include ``stsci.distutil``'s path.

- Added zest.releaser hooks for tweaking the development version string
  template, and for uploading new releases to STScI's local package index.


0.2.2 (2011-11-09)
------------------

- Fixed check for the issue205 bug on actual setuptools installs; before it
  only worked on distribute.  setuptools has the issue205 bug prior to version
  0.6c10.

- Improved the fix for the issue205 bug, especially on setuptools.
  setuptools, prior to 0.6c10, did not back of sys.modules either before

stsci/distutils/hooks.py
from __future__ import with_statement

import datetime
import glob
import os
import string
import sys

from distutils import log


try:
    from packaging.util import split_multiline
except ImportError:
    try:
        from distutils2.util import split_multiline
    except ImportError:
        from d2to1.util import split_multiline


try:
    # python 2
    reload
except NameError:
    # python 3
    from imp import reload

from .svnutils import get_svn_info, get_svn_version
from .versionutils import (package_uses_version_py, clean_version_py,
                           update_setup_datetime, VERSION_PY_TEMPLATE)

# For each version.py that we create, we will note the version of
# stsci.distutils (this package) that created it.  The problem is
# when the package being installed is stsci.distutils -- we don't
# know the version yet.  So, if we can't import the version (because
# it does not exist yet), we declare it to be None and special case
# it later.
try :
    from . import version as my_version
except ImportError :
    my_version = None


def is_display_option(ignore=None):
    """A hack to test if one of the arguments passed to setup.py is a display
    argument that should just display a value and exit.  If so, don't bother
    running this hook (this capability really ought to be included with
    distutils2).

    Optionally, ignore may contain a list of display options to ignore in this
    check.  Each option in the ignore list must contain the correct number of
    dashes.
    """

    from setuptools.dist import Distribution

    # If there were no arguments in argv (aside from the script name) then this
    # is an implied display opt
    if len(sys.argv) < 2:
        return True

    display_opts = ['--command-packages', '--help', '-h']

    for opt in Distribution.display_options:
        display_opts.append('--' + opt[0])

    for arg in sys.argv:
        if arg in display_opts and arg not in ignore:
            return True

    return False


# TODO: With luck this can go away soon--packaging now supports adding the cwd
# to sys.path for running setup_hooks.  But it also needs to support adding
# packages_root.  Also, it currently does not support adding cwd/packages_root
# to sys.path for pre/post-command hooks, so that needs to be fixed.
def use_packages_root(config):
    """
    Adds the path specified by the 'packages_root' option, or the current path
    if 'packages_root' is not specified, to sys.path.  This is particularly
    useful, for example, to run setup_hooks or add custom commands that are in
    your package's source tree.

    Use this when the root of your package source tree is not in
    the same directory with the setup.py

    Config Usage::

        [files]
        packages_root = lib
        # for installing pkgname from lib/pkgname/*.py

        [global]
        setup_hooks = stsci.distutils.hooks.use_packages_root

    """

    if 'files' in config and 'packages_root' in config['files']:
        root = config['files']['packages_root']
    else:
        root = ''

    if root not in sys.path:
        if root and sys.path[0] == '':
            sys.path.insert(1, root)
        else:
            sys.path.insert(0, root)

    # Reload the stsci namespace package in case any new paths can be added to
    # it from the new sys.path entry; depending on how the namespace packages
    # are installed this may fail--specifically if it's using the old
    # setuptools-based .pth approach there is not importable package called
    # 'stsci' that can be imported by itself.
    if 'stsci' in sys.modules:
        mod = sys.modules['stsci']
        if sys.version_info[:2] >= (3, 3) and not hasattr(mod, '__loader__'):
            # Workaround for Python bug #17099 on Python 3.3, where reload()
            # crashes if a module doesn't have an __loader__ attribute
            del sys.modules['stsci']
            try:
                import stsci
            except ImportError:
                pass
        else:
            try :
                reload(sys.modules['stsci'])
            except ImportError:
                # doesn't seem to bother anything when this reload() fails
                pass


def tag_svn_revision(config):
    """
    A setup_hook to add the SVN revision of the current working copy path to
    the package version string, but only if the version ends in .dev.

    For example, ``mypackage-1.0.dev`` becomes ``mypackage-1.0.dev1234``.  This
    is in accordance with the version string format standardized by PEP 386.

    This should be used as a replacement for the ``tag_svn_revision`` option to
    the egg_info command.  This hook is more compatible with
    packaging/distutils2, which does not include any VCS support.  This hook is
    also more flexible in that it turns the revision number on/off depending on
    the presence of ``.dev`` in the version string, so that it's not
    automatically added to the version in final releases.

    This hook does require the ``svnversion`` command to be available in order
    to work.  It does not examine the working copy metadata directly.


    Config Usage::

        [global]
        setup_hooks = stsci.distutils.hooks.tag_svn_revision

    You should write exactly this in your package's ``__init__.py``::

        from .version import *

    """

    if 'metadata' in config and 'version' in config['metadata']:
        metadata = config['metadata']
        version = metadata['version']

        # Don't add an svn revision unless the version ends with .dev
        if not version.endswith('.dev'):
            return

        # First try to get the revision by checking for it in an existing
        # .version module
        package_dir = config.get('files', {}).get('packages_root', '')
        packages = config.get('files', {}).get('packages', '')
        packages = split_multiline(packages)
        rev = None
        for package in packages:
            version_py = package_uses_version_py(package_dir, package)
            if not version_py:
                continue
            try:
                mod = __import__(package + '.version',
                                 fromlist='__svn_revision__')
            except ImportError:
                mod = None
            if mod is not None and hasattr(mod, '__svn_revision__'):
                rev = mod.__svn_revision__
                break

        # Cleanup
        names = set([package, package + '.'])
        for modname in list(sys.modules):
            if modname == package or modname.startswith(package + '.'):
                del sys.modules[modname]

        if rev is None:
            # A .version module didn't exist or was incomplete; try calling
            # svnversion directly
            rev = get_svn_version()

        if not rev:
            return
        if ':' in rev:
            rev, _ = rev.split(':', 1)
        while rev and rev[-1] not in string.digits:
            rev = rev[:-1]
        try:
            rev = int(rev)
        except (ValueError, TypeError):
            return

        metadata['version'] ='%s%d' % (version, rev)


def _version_hook(function_name, package_dir, packages, name, version, vdate):
    """This command hook creates an version.py file in each package that
    requires it.  This is by determining if the package's ``__init__.py`` tries
    to import or import from the version module.

    version.py will not be created in packages that don't use it.  It should
    only be used by the top-level package of the project.

    Don't use this function directly--instead use :func:`version_setup_hook` or
    :func:`version_pre_command_hook` which know how to retrieve the required
    metadata depending on the context they are run in.

    Not called directly from the config file.  See :func:`version_setup_hook`.

    """

    # Strip any revision info from version; that will be handled separately
    if '-' in version:
        version = version.split('-', 1)[0]

    for package in packages:
        version_py = package_uses_version_py(package_dir, package)
        if not version_py:
            continue

        rev = get_svn_version()
        if ((not rev or not rev[0] in string.digits) and
                os.path.exists(version_py)):
            # If were unable to determine an SVN revision and the version.py
            # already exists, just update the __setup_datetime__ and leave the
            # rest of the file untouched
            update_setup_datetime(version_py)
            return
        elif rev is None:
            rev = 'Unable to determine SVN revision'

        svn_info = get_svn_info()

        # Wrap version, rev, and svn_info in str() to ensure that Python 2
        # unicode literals aren't written, which will break things in Python 3
        template_variables = {
                'hook_function': function_name,
                'name': name,
                'version': str(version),
                'vdate': str(vdate),
                'svn_revision': str(rev),
                'svn_full_info': str(svn_info),
                'setup_datetime': datetime.datetime.now(),
        }

        # my_version is version.py for the stsci.distutils package.
        # It can be None if we are called during the install of
        # stsci.distutils; we are creating the version.py, so it was
        # not available to import yet.  If this is what is happening,
        # we construct it specially.
        if my_version is None :
            if  package == 'stsci.distutils' :
                template_variables['stsci_distutils_version'] = version
            else:
                # It should never happen that version.py does not
                # exist when we are installing any other package.
                raise RuntimeError('Internal consistency error')
        else :
            template_variables['stsci_distutils_version'] = \
                    my_version.__version__

        with open(version_py, 'w') as f:
            f.write(VERSION_PY_TEMPLATE % template_variables)


def version_setup_hook(config):
    """Creates a Python module called version.py which contains these variables:

    * ``__version__`` (the release version)
    * ``__svn_revision__`` (the SVN revision info as returned by the ``svnversion``
      command)
    * ``__svn_full_info__`` (as returned by the ``svn info`` command)
    * ``__setup_datetime__`` (the date and time that setup.py was last run).
    * ``__vdate__`` (the release date)

    These variables can be imported in the package's ``__init__.py`` for
    degugging purposes.  The version.py module will *only* be created in a
    package that imports from the version module in its ``__init__.py``.  It
    should be noted that this is generally preferable to writing these
    variables directly into ``__init__.py``, since this provides more control
    and is less likely to unexpectedly break things in ``__init__.py``.

    Config Usage::

        [global]
        setup-hooks = stsci.distutils.hooks.version_setup_hook

    You should write exactly this in your package's ``__init__.py``::

        from .version import *

    """

    if is_display_option(ignore=['--version']):
        return

    name = config['metadata'].get('name')
    version = config['metadata'].get('version', '0.0.0')
    vdate = config['metadata'].get('vdate', 'unspecified')
    package_dir = config.get('files', {}).get('packages_root', '')
    packages = config.get('files', {}).get('packages', '')

    packages = split_multiline(packages)

    _version_hook(__name__ + '.version_setup_hook', package_dir, packages,
                 name, version, vdate)


def version_pre_command_hook(command_obj):
    """
    .. deprecated:: 0.3
        Use :func:`version_setup_hook` instead; it's generally safer to
        check/update the version.py module on every setup.py run instead of on
        specific commands.

    This command hook creates an version.py file in each package that requires
    it.  This is by determining if the package's ``__init__.py`` tries to
    import or import from the version module.

    version.py will not be created in packages that don't use it.  It should
    only be used by the top-level package of the project.
    """

    if is_display_option():
        return

    package_dir = command_obj.distribution.package_dir.get('', '.')
    packages = command_obj.distribution.packages
    name = command_obj.distribution.metadata.name
    version = command_obj.distribution.metadata.version

    _version_hook(__name__ + '.version_pre_command_hook',package_dir, packages,
                 name, version, vdate=None)


def version_post_command_hook(command_obj):
    """
    .. deprecated:: 0.3
        This hook was meant to complement :func:`version_pre_command_hook`,
        also deprecated.

    Cleans up a previously generated version.py in order to avoid
    clutter.

    Only removes the file if we're in an SVN working copy and the file is not
    already under version control.
    """

    package_dir = command_obj.distribution.package_dir.get('', '.')
    packages = command_obj.distribution.packages

    for package in packages:
        clean_version_py(package_dir, package)


def numpy_extension_hook(command_obj):
    """A distutils2 pre-command hook for the build_ext command needed for
    building extension modules that use NumPy.

    To use this hook, add 'numpy' to the list of include_dirs in setup.cfg
    section for an extension module.  This hook will replace 'numpy' with the
    necessary numpy header paths in the include_dirs option for that extension.

    Note: Although this function uses numpy, stsci.distutils does not depend on
    numpy.  It is up to the distribution that uses this hook to require numpy
    as a dependency.

    Config Usage::

        [extension=mypackage.extmod]
        sources =
            foo.c
            bar.c
        include_dirs = numpy

        [build_ext]
        pre-hook.numpy-extension = stsci.distutils.hooks.numpy_extension_hook
    """

    command_name = command_obj.get_command_name()
    if command_name != 'build_ext':
        log.warn('%s is meant to be used with the build_ext command only; '
                 'it is not for use with the %s command.' %
                 (__name__, command_name))
    try:
        import numpy
    except ImportError:
        # It's virtually impossible to automatically install numpy through
        # setuptools; I've tried.  It's not pretty.
        # Besides, we don't want users complaining that our software doesn't
        # work just because numpy didn't build on their system.
        sys.stderr.write('\n\nNumpy is required to build this package.\n'
                         'Please install Numpy on your system first.\n\n')
        sys.exit(1)

    #includes = [numpy.get_include()]
    includes = [numpy.get_numarray_include(), numpy.get_include()]
    for extension in command_obj.extensions:
        if 'numpy' not in extension.include_dirs:
            continue
        idx = extension.include_dirs.index('numpy')
        for inc in includes:
            extension.include_dirs.insert(idx, inc)
        extension.include_dirs.remove('numpy')


def glob_data_files(command_obj):
    """A pre-command hook for the install_data command allowing wildcard
    patterns to be used in the data_files option.

    Also ensures that data files with relative paths as their targets are
    installed relative install_lib.

    Config Usage::

        [files]
        data_files =
            target_directory = source_directory/*.foo
            other_target_directory = other_source_directory/*

        [install_data]
        pre-hook.glob-data-files = stsci.distutils.hooks.glob_data_files


    """

    command_name = command_obj.get_command_name()
    if command_name != 'install_data':
        log.warn('%s is meant to be used with the install_data command only; '
                 'it is not for use with the %s command.' %
                 (__name__, command_name))

    data_files = command_obj.data_files

    for idx, val in enumerate(data_files[:]):
        if isinstance(val, basestring):
            # Support the rare, deprecated case where just a filename is given
            filenames = glob.glob(val)
            del data_files[idx]
            data_files.extend(filenames)
            continue

        dest, filenames = val
        filenames = sum((glob.glob(item) for item in filenames), [])
        data_files[idx] = (dest, filenames)

    # Ensure the correct install dir; this is the default behavior for
    # installing with distribute, but when using
    # --single-version-externally-managed we need to to tweak this
    install_cmd = command_obj.get_finalized_command('install')
    if command_obj.install_dir == install_cmd.install_data:
        install_lib_cmd = command_obj.get_finalized_command('install_lib')
        command_obj.install_dir = install_lib_cmd.install_dir
